lib/grammers-client/src/client/chats.rs

```diff
@@ -725,6 +725,55 @@ impl Client {
         let permissions = ParticipantPermissions::Channel(participant.participant);
         Ok(permissions)
     }
+    
+    /// join private chat
+    pub async fn accept_invite_link(
+        &mut self,
+        invite_link: &str,
+    ) -> Result<Option<Chat>, InvocationError> {
+        use tl::enums::Updates;
+        assert!(invite_link.starts_with("https://t.me/joinchat/"));
+        let update_chat = match self
+            .invoke(&tl::functions::messages::ImportChatInvite {
+                hash: invite_link.replace("https://t.me/joinchat/", ""),
```

Comment:
This is very brittle. There is a lot more invite links Telegram supports (for example, the `https` could be `http`, `t.me` could be `telegram.me` or even `telegram.dog`, and `joinchat/` could be just `+` among others).

This code should properly validate the invite link and extract the join hash, rather than simply replacing it away.

Reply:
Here I can think of the way to deal with it is to introduce regex, is there any other better way?

Reply:
We have two options:

* Let the user deal with it. The user should provide the hash, not a link.
* Parse the URL. We can write a "parser" for the few supported URLs, or add a new dependency on a crate capable of parsing URLs. Once the URL is parsed, validate the domain and extract the hash from the path.

I would prefer to avoid depending on a regex crate if possible.

Reply:
Well, when I use option two to modify (the `url` crate used), I almost crash. Although I think the checking of the rules is done, when I want to get the hash in the URL, I find that a lot of replacements are needed. Maybe my way is too simplistic
```rust
use grammers_mtproto::mtp::RpcError;
        let url_parse = url::Url::parse(invite_link).map_err(|_| {
            InvocationError::Rpc(RpcError {
                code: 400,
                name: "INVITE_HASH_INVALID".to_string(),
                value: None,
                caused_by: None,
            })
        })?;

        let (scheme, host, path) = (url_parse.scheme(), url_parse.host_str(), url_parse.path());

        if !vec!["https", "http"].contains(&scheme)
            || host.is_none()
            || (!path.starts_with("joinchat/") || !path.starts_with("+"))
        {
            return Err(InvocationError::Rpc(RpcError {
                code: 400,
                name: "INVITE_HASH_INVALID".to_string(),
                value: None,
                caused_by: None,
            }));
        }

        let host = host.unwrap();
        let hosts = host.split(".").collect::<Vec<&str>>();
        let host_check_result = match hosts.len() {
            2 if vec!["t", "telegram"].contains(&hosts[0])
                && vec!["me", "org", "dog"].contains(&hosts[0]) =>
            {
                true
            }
            3 if "www" == hosts[0]
                && vec!["t", "telegram"].contains(&hosts[1])
                && vec!["me", "org", "dog"].contains(&hosts[2]) =>
            {
                true
            }
            _ => false,
        };

        if !host_check_result {
            return Err(InvocationError::Rpc(RpcError {
                code: 400,
                name: "INVITE_HASH_INVALID".to_string(),
                value: None,
                caused_by: None,
            }));
        }
```

Reply:
Both option 1 and option 2 are supported